### PR TITLE
fix order of cxf handlers to enable symmetric tracing around jaxws handler chain

### PR DESCRIPTION
**Pull request by pellmont**
the current implementation of Start and End around the invocation of a Jax WS is asymmetric around the JAX-WS Handler Chain.

Current behavior:
(execution of incoming  MessageHandlers) -> (TracingStartInInterceptor) -> (WebService Invocation) -> (execution of outgoing MessageHandlers) -> (TracingEndInInterceptor)

if I  understood the code of this cxf instrumentation correctly, the intent was to build the span close around the WebService Invocation (without Handler Chains).

So the desired behavior would look like this:
(execution of incoming  MessageHandlers) -> (TracingStartInInterceptor) -> (WebService Invocation) -> (TracingEndInInterceptor) -> (execution of outgoing MessageHandlers)

Unfortunately CXF is calling the Outgoing Chain inside the POST_INVOKE Phase of Cxf (so the outgoing chain is technically a sub-chain in the incoming chain... which is documented but quite surprising...).

So the solution in the fix at least guarantees the the outgoing chain is invoked AFTER end of tracing. For any extra Interceptors in the POST_INVOKE Phase there is still no guarantee of  ordering, but I think this is not a opentelemetry issue but a design-flaw of CXF...

### CODE REVIEW

**Comment by laurit** [comment]
I think we should also modify tests. Perhaps add
```groovy
    endpoint.getOutInterceptors().add(new AbstractPhaseInterceptor<Message>(Phase.SEND) {
      @Override
      void handleMessage(Message message) throws Fault {
        Context context = Context.current()
        if (LocalRootSpan.fromContext(context) != Span.fromContext(context)) {
          throw new IllegalStateException("handler span should not be ended before outgoing interceptors")
        }
      }
    })
```
after https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/aceb5b2ee9f5b97abcf8599cdb5d2007eb4dad49/instrumentation/jaxws/jaxws-2.0-cxf-3.0/javaagent/src/test/groovy/TestWsServlet.groovy#L20

**Review by laurit — APPROVED** [review]

File: instrumentation/jaxws/jaxws-2.0-cxf-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cxf/TracingEndInInterceptor.java [hunk 1]
```diff
@@ -5,13 +5,15 @@
...
 
+import org.apache.cxf.interceptor.OutgoingChainInterceptor;
 import org.apache.cxf.message.Message;
 import org.apache.cxf.phase.AbstractPhaseInterceptor;
 import org.apache.cxf.phase.Phase;
 
 public class TracingEndInInterceptor extends AbstractPhaseInterceptor<Message> {
   public TracingEndInInterceptor() {
     super(Phase.POST_INVOKE);
+    addBefore(OutgoingChainInterceptor.class.getName());
```
Comment (trask) [inline instrumentation/jaxws/jaxws-2.0-cxf-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cxf/TracingEndInInterceptor.java:16]:
```suggestion
    // end the span before the OutInterceptors (callbacks) are executed
    addBefore(OutgoingChainInterceptor.class.getName());
```

**Review by trask — APPROVED** [review]
thanks @pellmont!

can you also update the test as @laurit suggested?

File: instrumentation/jaxws/jaxws-2.0-cxf-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cxf/TracingEndInInterceptor.java (hunk 1 above)
Comment (pellmont) [reply instrumentation/jaxws/jaxws-2.0-cxf-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cxf/TracingEndInInterceptor.java:17]: Thanks for the review.
Sorry for the missing test, I just added it.

**Pull request merged**